### ISSUE TITLE
template engine does not work to generate code

### ISSUE THREAD

**Issue by despiegk** [vlang/v#13195]
my plan was to generate code using vlang templates, which would be awesome, but I can't because of { and }

```
module data
import texttools

/////////// THIS IS THE TEMPLATE, THIS CAN BE MODIFIED

///////IMPORTANT DO NOT MODIFY THIS FILE, THIS FILE IS GENERATED FROM 'data_obj_template'
///////JUST REMOVE THIS FILE IF YOU WANT IT TO BE REGENERATED

//add, id can be the name
pub fn @object_new(id string) ?&@object_u {
	mut id2 := id
	id2 = texttools.name_fix(id)
	mut df := factory()
	if id2 in df.@objects{
		return df.@objects[id2]
	}
	mut obj := &@object_u{id:id2}
	df.@objects[id2] = obj
}

//get, if not found will give error
pub fn @object_get(id string) ?&@object_u {
	mut df := factory()
	mut id2 := id
	id2 = texttools.name_fix(id)
	if id2 in df.@objects{
		return df.@objects[id2]
	}
	return error("cannot find @object @id2")
}

//check the data obj exists
pub fn @object_exists(id string) bool {
	mut df := factory()
	mut id2 := id
	id2 = texttools.name_fix(id)	
	return id2 in df.@objects
}

```

code which generated
```
module main
import os

const codepath_file = @FILE


fn  main() {
	force := true
	togenerate := ["user","circle"]
	dirpath := os.dir(codepath_file)
	println(dirpath)
	for name in togenerate{
		methodfilepath := "${dirpath}/${name}_methods.v"
		if os.exists(methodfilepath) && ! force{
			continue
		}
		object :=  name
		object_u := name.capitalize()
		objects :=  name+"s"
		objects_u := name.capitalize()+"s"
		mut txt := $tmpl('data_obj.v.templ')
		txt = txt.replace("/////////// THIS IS THE TEMPLATE, THIS CAN BE MODIFIED","")
		println(txt)
		os.write_file(methodfilepath,txt) or {panic("canont write file $methodfilepath")}
	}
	//data_obj.v.templ
}


```

result is

```
module data

//Data object for User
pub struct User {
pub mut:
	id          int
	name        string
	description string
	tags		[]string
	remarks		[]int
</div>
```

it stops with the div

**Comment by ArtemkaKun** [comment]
Hi @despiegk, I'm not a pro in V templates, but afaik templates only work for doing templates for web stuff (HTML, CSS, JS).
Why it inserts </dev> - this is because template parsers defines type of context block and HTML is default.

You should probably confirm these 2 things with V-dev in Discord or Telegram:
1. if there is the possibility of template usage in V (because if I recall right some kind of meta-programming [I forgot the name of it], which is allowed in C/C++, isn't allowed in V because it creates complex code)
2. if the first step is true -> confirm is there is parser of templates for V code, not for web stuff

**Comment by spytheman** [comment]
I can not reproduce the problem 🤔 , I get:
```
0[10:17:42]delian@nemesis: (master) /v/vnew/bug $ v run .
internal_memory:18:18: warning: use `${f(expr)}` instead of `$f(expr)`
internal_memory:30:18: warning: use `${f(expr)}` instead of `$f(expr)`
internal_memory:41:21: warning: use `${f(expr)}` instead of `$f(expr)`
:18:20: error: unexpected string ` string) ?&`, expecting `,`
1[10:17:44]delian@nemesis: (master) /v/vnew/bug $ ls -lart 
total 16
-rw-rw-r--  1 delian delian  656 Jan 17 10:11 main.v
-rw-rw-r--  1 delian delian  894 Jan 17 10:12 data_obj.v.templ
drwxrwxr-x 13 delian delian 4096 Jan 17 10:17 ..
drwxrwxr-x  2 delian delian 4096 Jan 17 10:17 .
0[10:17:49]delian@nemesis: (master) /v/vnew/bug $
```

**Comment by spytheman** [comment]
Imho the whole template subsystem needs to be rethought and re-implemented with a proper parser, to avoid special/edge cases and to have more useful error reporting.

**Comment by despiegk** [comment]
a simple template engine should work IMO, ability to say not to support {} should already do a lot
there are 2 templates, the one above indeed gives this issue, which is also an issue

```
module data

//Data object for @Object
pub struct @Object {
pub mut:
	id          int
	name        string
	description string
	tags		[]string
	remarks		[]int
}
```

this was the other template

**Comment by spytheman** [comment]
> this was the other template

thanks

**Comment by spytheman** [comment]
> a simple template engine should work IMO, ability to say not to support {} should already do a lot

I totally agree that it should work, and it should be more robust.

**Comment by spytheman** [comment]
> if there is the possibility of template usage in V (because if I recall right some kind of meta-programming [I forgot the name of it], which is allowed in C/C++, isn't allowed in V because it creates complex code)

The text templates should be usable for arbitrary text generation, including V source code, that then the user can use for whatever he wants - storing in a DB, printing, saving in a file, etc. The compilation of the template + the variables from the calling code, happens at compiletime, but the substitution happens at runtime, which is totally fine.

What V does not support, is C++ templates and C style macros.

**Comment by spytheman** [comment]
@despiegk the problem should be fixed in V 40a5c5c.

 I've changed the parser for `$tmpl` to use the older more complex html mode, only for template files, that have the .html extension.

All other template files, will be parsed in a simpler way, that just supports `@variable`, `@include`, `@if`, `@else`, `@for`, `@end` , with no additional heuristics, so they should be more robust.

Please check, if that solves the issue for you.

**Comment by medvednikov** [comment]
Fixed.
